src/charm.py

```diff
@@ -87,6 +92,115 @@ def _spring_boot_port(self) -> int:
+        """Parse numbers with human-readable units like K, M and G.
+
+        Args:
+            number: input string, something like ``"1G"`` or ``"33m"``.
+
+        Returns:
+            Parsed result, as an integer.
+
+        Raises:
+            ValueError: when the input number is invalid.
+        """
+        number = number.lower()
+        unit = number[-1].lower()
+        unit_scale = {"k": 2**10, "m": 2**20, "g": 2**30}
+        if unit not in unit_scale:
+            try:
+                return int(number)
+            except ValueError as exc:
+                raise ValueError(f"Unknown human-readable unit: {repr(number)}") from exc
+        digits = number[:-1]
+        return int(digits) * unit_scale[unit]
+
+    def _extract_jvm_heap_initial_memory(self, jvm_config: str) -> int:
+        """Get the JVM heap initial memory size from a JVM configuration string.
+
+        Args:
+            jvm_config: JVM configuration string.
+
+        Returns:
+            JVM heap initial memory size in number of bytes. ``0`` if not defined in the JVM
+            configuration string.
+        """
+        candidates = re.findall("(^|\\s)(-Xms\\d+[kmgtKMGT]?)\\b", jvm_config)
+        if not candidates:
+            return 0
+        return self._parse_human_readable_units(candidates[-1][-1].removeprefix("-Xms"))
+
+    def _extract_jvm_heap_maximum_memory(self, jvm_config: str) -> int:
+        """Get the JVM heap maximum memory size from a JVM configuration string.
+
+        Args:
+            jvm_config: JVM configuration string.
+
+        Returns:
+            JVM heap maximum memory size in number of bytes. ``0`` if not defined in the JVM
+            configuration string.
+        """
+        candidates = re.findall("(^|\\s)(-Xmx\\d+[kmgtKMGT]?)\\b", jvm_config)
+        if not candidates:
+            return 0
+        return self._parse_human_readable_units(candidates[-1][-1].removeprefix("-Xmx"))
+
+    def _jvm_config(self) -> str:
+        """Get and verify the JVM parameters defined in the charm configuration jvm-config.
+
+        Returns:
+            JVM command line arguments as a string.
+
+        Raises:
+            ReconciliationError: when the jvm-config value is invalid.
+        """
+        config = self.model.config["jvm-config"]
+        if not config:
+            return ""
+        java_heap_initial_memory = self._extract_jvm_heap_initial_memory(config)
+        java_heap_maximum_memory = self._extract_jvm_heap_maximum_memory(config)
+        container_memory_limit = self._get_sprint_boot_container_memory_constraint()
+        if (
+            container_memory_limit
+            and max(java_heap_maximum_memory, java_heap_initial_memory) > container_memory_limit
+        ):
+            raise ReconciliationError(
+                new_status=BlockedStatus(
+                    "Invalid jvm-config, "
+                    "Java heap memory requirement exceeds application memory constraint"
```

Comment:
Could this be misleading if it is triggered by the `java_heap_initial_memory` being larger than the memory limit?

Reply:
I updated the word `requirement` to `specification`, thanks!